src/controller/c/supervisor.c

```diff
@@ -2056,6 +2083,84 @@ const char *wb_supervisor_node_get_base_type_name(WbNodeRef node) {
   return wb_node_get_name(node->type);
 }
 
+WbFieldRef wb_supervisor_node_get_field_by_index(WbNodeRef node, int index) {
+  if (!robot_check_supervisor(__FUNCTION__))
+    return NULL;
+
+  if (!is_node_ref_valid(node)) {
+    if (!robot_is_quitting())
+      fprintf(stderr, "Error: %s() called with a NULL or invalid 'node' argument.\n", __FUNCTION__);
+    return NULL;
+  }
+  const int node_fields = wb_supervisor_node_get_number_of_fields(node);
+  if (index < 0) {
+    if (!robot_is_quitting())
+      fprintf(stderr, "Error: %s() called with an invalid 'index' argument (%d while node has %d fields).\n", __FUNCTION__,
+              index, node_fields);
+    return NULL;
+  }
+
+  robot_mutex_lock_step();
+  // search if field is already present in field_list
+  WbFieldRef result = find_field_by_index(index, node->id);
+  if (!result) {
+    // otherwise: need to talk to Webots
+    WbFieldRef field_list_before = field_list;
+    requested_field_index = index;
+    node_ref = node->id;
+    wb_robot_flush_unlocked();
+    requested_field_index = -1;
+    if (field_list != field_list_before)
+      result = field_list;
+    else
+      result = find_field_by_index(index, node->id);
+    if (result && node->is_proto_internal)
+      result->is_proto_internal = true;
+  }
+  robot_mutex_unlock_step();
+  return result;
+}
+
+WbFieldRef wb_supervisor_node_get_proto_field_by_index(WbNodeRef node, int index) {
+  if (!robot_check_supervisor(__FUNCTION__))
+    return NULL;
+
+  if (!is_node_ref_valid(node)) {
+    if (!robot_is_quitting())
+      fprintf(stderr, "Error: %s() called with a NULL or invalid 'node' argument.\n", __FUNCTION__);
+    return NULL;
+  }
+  const int node_fields = wb_supervisor_node_get_proto_number_of_fields(node);
```

Comment:
You also have to remove this line given that `node_fields` is no longer used and adjust the error message.